[SPARK-9851] Support submitting map stages individually in DAGScheduler

mateiz opened the pull request:
This patch adds support for submitting map stages in a DAG individually so that we can make downstream decisions after seeing statistics about their output, as part of SPARK-9850. I also added more comments to many of the key classes in DAGScheduler. By itself, the patch is not super useful except maybe to switch between a shuffle and broadcast join, but with the other subtasks of SPARK-9850 we'll be able to do more interesting decisions.

The main entry point is SparkContext.submitMapStage, which lets you run a map stage and see stats about the map output sizes. Other stats could also be collected through accumulators. See AdaptiveSchedulingSuite for a short example.

mateiz committed:
Initial work on adaptive execution in DAGScheduler

mateiz committed:
Remove reduce task stuff, clean up, add docs and tests

mateiz committed:
Add a test to FailureSuite too

mateiz committed:
Another test suite

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40822 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40822/consoleFull) for   PR 8180 at commit [`5092c54`](https://github.com/apache/spark/commit/5092c54436cd6b060ebd3c11d49ff709985bf8cb).

AmplabJenkins commented:
Merged build finished. Test FAILed.

mateiz commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40829 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40829/consoleFull) for   PR 8180 at commit [`732c4ed`](https://github.com/apache/spark/commit/732c4eda96298a9d45400807d919286edbc7ba5d).

SparkQA commented:
[Test build #40822 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40822/console) for   PR 8180 at commit [`5092c54`](https://github.com/apache/spark/commit/5092c54436cd6b060ebd3c11d49ff709985bf8cb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #40829 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40829/console) for   PR 8180 at commit [`732c4ed`](https://github.com/apache/spark/commit/732c4eda96298a9d45400807d919286edbc7ba5d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

mateiz commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40848 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40848/consoleFull) for   PR 8180 at commit [`732c4ed`](https://github.com/apache/spark/commit/732c4eda96298a9d45400807d919286edbc7ba5d).

SparkQA commented:
[Test build #40848 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40848/console) for   PR 8180 at commit [`732c4ed`](https://github.com/apache/spark/commit/732c4eda96298a9d45400807d919286edbc7ba5d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40926 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40926/consoleFull) for   PR 8180 at commit [`d301862`](https://github.com/apache/spark/commit/d301862e6ca3cfda401f8b84a8e0000915f55c7c).

SparkQA commented:
[Test build #40926 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40926/console) for   PR 8180 at commit [`d301862`](https://github.com/apache/spark/commit/d301862e6ca3cfda401f8b84a8e0000915f55c7c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class StringIndexerModel (`
  * `  implicit class StringToColumn(val sc: StringContext) `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40926/
Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41285 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41285/consoleFull) for   PR 8180 at commit [`4a47919`](https://github.com/apache/spark/commit/4a479199bb7bb54403f2a112463817c4246b4e35).

SparkQA commented:
[Test build #41285 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41285/console) for   PR 8180 at commit [`4a47919`](https://github.com/apache/spark/commit/4a479199bb7bb54403f2a112463817c4246b4e35).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PromotePrecision(child: Expression) extends UnaryExpression `
  * `case class CheckOverflow(child: Expression, dataType: DecimalType) extends UnaryExpression `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41285/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41286/consoleFull) for   PR 8180 at commit [`210ca43`](https://github.com/apache/spark/commit/210ca431ec5e0d51c0e206a5eca549851c470e5e).

SparkQA commented:
[Test build #41286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41286/console) for   PR 8180 at commit [`210ca43`](https://github.com/apache/spark/commit/210ca431ec5e0d51c0e206a5eca549851c470e5e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41286/
Test PASSed.

rxin commented:
@zsxwing if you have time, can you review this pr?

squito commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -47,15 +47,48 @@ import org.apache.spark.storage.BlockManagerMessages.BlockManagerHeartbeat
+ *    boundaries, which introduce a barrier (where we must wait for the previous stage to finish to
+ *    fetch outputs). There are two types of stages: [[ResultStage]], for the final stage that
+ *    executes an action, and [[ShuffleMapStage]], which writes map output files for a shuffle.
+ *    Stages are often shared across multiple jobs, if these jobs reuse the same RDDs.
```
Its nice to see these expanded comments, but I think we really need to add a section on stage attempts.  That is most probably the most confusing part of the dag schduler and where most bugs occur.

squito commented on core/src/main/scala/org/apache/spark/scheduler/ShuffleMapStage.scala:
```diff
@@ -37,6 +45,9 @@ private[spark] class ShuffleMapStage(
 
   override def toString: String = "ShuffleMapStage " + id
 
+  /** Map-stage jobs that were submitted to execute this stage independently (if any) */
```
can you expand that this only for jobs that that were submitted *just* for this stage?  Ie., it does *not* include jobs that indirectly depend on this stage. 
Also, this is kinda obvious from the type, but do you mind expanding the comment to say that this only *running* jobs that were submitted for this job?  So that if  job completes, its removed from this list.

squito commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -812,7 +971,7 @@ class DAGScheduler(
           val job = s.resultOfJob.get
           partitionsToCompute.map { id =>
-            val p = job.partitions(id)
+            val p = s.partitions(id)
```
I've wondered about this line many times in the past, but been too scared to change it -- is there any reason why `job.partitions` will be different from `resultStage.partitions`?

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -990,6 +999,110 @@ class DAGSchedulerSuite
+    assert(results === Map(0 -> 44, 1 -> 45))
+    results.clear()
+    assertDataStructuresEmpty()
+  }
```
I think we'll need a bunch more cases around stage failure, as this has been one of the most error prone parts of the scheduler.  Things like:
* direct execution of map stage, after the stage has already failed from a result job that depends on it (and vice versa)
* map stage retry with zombie task completion, so we get double completion of shuffle map tasks
* map state retry with zombie task completion, so map output is made available before any single task set has completed.  (This doesn't currently work for normal jobs either, see SPARK-5259 -- but after that fix is in, would be good to make sure it works with this case as well.)
* map stage submission, which in turn depends on a lineage of more than one shuffle dependency.  failure in one of the fetches should result in the proper set of retries
* map stage submission with a fetch failure, which results in a retry of a "skipped" stage.  (this is overdue a test case for regular job submission in any case -- I have been meaning to do this.  I can look at this and try to give you a test case for map stage submission as well.)

squito commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -746,6 +848,63 @@ class DAGScheduler(
+        listener.jobFailed(e)
+        return
+    }
+    if (finalStage != null) {
```
can you add a comment on how `finalStage` can be `null`?  would be nice to also have this in `handleJobSubmitted` as well, actually

mateiz replied:
Looks like it actually can't be null due to the return above, so I'll remove this. I think it was put in when someone refactored the code.

squito commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1973,6 +1973,24 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+  }
+
+
+
```
nit: too many spaces

squito commented on core/src/main/scala/org/apache/spark/MapOutputStatistics.scala:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Holds statistics about the output sizes in a map stage. May become a DeveloperApi in the future.
+ */
+private[spark] class MapOutputStatistics(val shuffleId: Int, val bytesByPartitionId: Array[Long])
```
can you add a comment to make it clear that `bytesByPartitionId` is indexed by the shuffle-*read* aka reduce side?  (at least, I think it is, if I'm reading the code right.)  Also probably worth adding a comment that this is just an estimate of the size, and can really be off by an arbitrary amount given the approximations made in `HighlyCompressedMapStatus`.

squito commented:
I'm *super* nervous about adding this, given our current inability to deal w/ the complexity of the scheduler.

markhamstra commented:
@squito Agree that this is high risk, but SPARK-9851 is also very high reward.  On balance, I'm excited about this, but also expecting some difficult debugging.

squito commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -132,13 +133,46 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
+   * @param shuffleId ID of the shuffle
+   * @param numPartitions number of map output partitions
+   */
+  def getStatistics(shuffleId: Int, numPartitions: Int): MapOutputStatistics = {
```
its a little strange that you need to pass `numPartitions` in here, since that is a property of the map output.  Maybe instead of just taking  `shuffleId`, it should take a `shuffleHandle`, so you can get the number of map output partitions w/ `shuffleHandle.dependency.partioner.numPartitions`?

mateiz replied:
Good point. It's a bit weird that we don't remember numPartitions anywhere else but it would be clearer this way.

mateiz commented:
Hey Imran, I'm curious, have you actually worked on stuff in the scheduler? I don't know what you mean about inability to deal with complexity in it, but it has gotten simpler over time (e.g. eliminating the ability for jobs to execute locally). If you look in this patch, it's only 100-200 lines of changes to the scheduler; the rest is docs and tests. And as a whole, the DAGScheduler is only around 1600 lines.

squito commented:
I’ve only recently looked at making changes to the scheduler, but it seems to me there is widespread agreement among committers that it is very error prone.  For example, consider Andrew Or’s plea in [SPARK-8987](https://issues.apache.org/jira/browse/SPARK-8987) which starts with:
> DAGScheduler is one of the most monstrous piece of code in Spark.

Other recent examples of similar sentiments are in this [dicussion on backporting SPARK-8103](https://github.com/apache/spark/pull/7572) or confusion in the earlier versions of SPARK-5945, SPARK-7308, and SPARK-8103.  Even this [seemingly innocuous three line change](https://github.com/apache/spark/commit/702aa9d7fb16c98a50e046edfd76b8a7861d0391#diff-6a9ff7fb74fd490a50462d45db2d5e11R792) inadvertently introduced SPARK-9809 (its really lucky that somebody stumbled on that before the release).

I’ve been working on issues related to fault-tolerance, primarily SPARK-8103 & SPARK-8029, which came from real customer escalations.  Those took me a *long* time to wrap my head around, after painfully trying to make sense of user logs, create a reproduction, propose a fix, convince others there really was something wrong, and get lots of help to make the right fix.  

I did a bit of fault-injection testing as well, and things seemed to pass consistently after my fixes, so I was hoping that would be the end of the story.  But then I dug through some existing jiras, and found [SPARK-5259](https://issues.apache.org/jira/browse/SPARK-5259).  I couldn't believe it had been open since January!  A community member had discovered it and even very clearly described exactly how it happened, but still we haven't fixed it.  We're about to release spark 1.5 with it still broken, which means that's at least 3 releases where fault tolerance is knowingly broken.  I find that embarrassing.

I'm not saying all of this to denigrate the effort that everyone has already put into it, but I just want to be clear that I really do mean it: we are unable to deal with the complexity of the scheduler.  IMO, the highest reward would be to fix the fault-tolerance issues, and focus on testing the scheduler so we gain more confidence in it.  So while this feature is interesting, I think we should proceed very cautiously.

squito commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -990,6 +999,110 @@ class DAGSchedulerSuite
+    // Ask the scheduler to try it again; TaskSet 2 will rerun the map task that we couldn't fetch
+    // from, then TaskSet 3 will run the reduce stage
+    scheduler.resubmitFailedStages()
+    complete(taskSets(2), Seq((Success, makeMapStatus("hostA", reduceRdd.partitions.size))))
```
as part of the SPARK-5945, @ilganeli is introducing a few util methods that make this a little more clear, eg. [`completeNextShuffleMapSuccessfully`](https://github.com/apache/spark/pull/5636/files#diff-f3b410b16818d8f34bb1eb4120a60d51R516).  IMO that helps make these test cases a bit more readable -- right now its a bit hard for the reader to know what `taskSets(2)` is.  Also if someone makes a change that breaks things, its much clearer if they can see eg. that they never launched attempt 1 for stage 0, rather than just getting some strange behavior later on.

mateiz committed:
tweaks

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

andrewor14 commented:
>  For example, consider Andrew Or’s plea in SPARK-8987 which starts with [...]

OK, my quote was taken out of its original context. It was intended to motivate the increase in test coverage in the scheduler, not to prevent additional changes made to it. This patch adds a lot of tests for the functionality it introduces so that's not a problem. I would say my issue is mostly unrelated to this patch.

SparkQA commented:
[Test build #41947 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41947/consoleFull) for   PR 8180 at commit [`dac96b7`](https://github.com/apache/spark/commit/dac96b7731ae51f44ee477a6e758f7b7131b3d2c).

SparkQA commented:
[Test build #41947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41947/console) for   PR 8180 at commit [`dac96b7`](https://github.com/apache/spark/commit/dac96b7731ae51f44ee477a6e758f7b7131b3d2c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41947/
Test PASSed.

srowen commented:
@andrewor14 listening from the sidelines, I think @squito was only pointing out that you have also noted how hard it is to change the scheduler with confidence. I don't think he meant to or did imply you are against this change; clearly the linked JIRA shows you are just favoring more tests as a remedy to the problem you describe.

There does seem to be some disconnect between SPARK-8987 and @mateiz's view that there's not much problem changing the scheduler. FWIW it does strike me as a big change to code that has been a source of subtle bugs in recent memory. So, maybe a little more nervousness is called for.

We can't stop changing the scheduler and I'm sure that the motivations for this change are worthy. What's the best way forward -- @squito pointed out some aspects that make him most nervous; would a few more tests help de-stress? I tend to respect his nervousness as he takes the brunt of these problems from big deployments, and getting this working for large deployments reliably is of course something everyone needs.

mateiz commented:
Before deciding whether it's a big change, do also take a look at the change. As I said, it's only about 100-200 lines of actual changes, the rest is comments for the existing code and tests. And the actual mechanism by which this works already existed in the DAGScheduler design (the scheduler already tracked map stages, including sharing them between jobs, etc, but there was no way to submit a map stage by itself). If you want to make it seem less scary, I can send a PR without the comments in the original scheduler :).

By the way, we have a working prototype of SPARK-9850 end-to-end (including prototypes of the other JIRAs in it) so it's not like this is untested either.

mateiz commented:
BTW, I'm working on updating this with a few more tests as suggested as well.

squito commented:
Sorry Andrew, I certainly didn't mean to imply that I knew your opinion on this particular patch -- as Sean said, I was just trying to point out the general situation.  My point is just that new features add complexity, and make it harder to fix bugs; and we have plenty of complexity and bugs now.  I don't mean to block the patch, its a cool feature.  I'm just nervous about any change to the dag scheduler (eg., even my own proposal in https://github.com/apache/spark/pull/8427, which is 30 lines + 100 lines of tests, and most likely needs more testing still).  Perhaps I err too much on the side of caution, I'm just providing a counterpoint.

Thanks for adding the the additional tests, Matei.  Btw, there is an example test for skipped stages you should be able to copy more or less here: https://github.com/apache/spark/pull/8402

mateiz committed:
A few more tests and docs (still want to add more)

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42004 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42004/consoleFull) for   PR 8180 at commit [`bb5190f`](https://github.com/apache/spark/commit/bb5190f0804f83fd178960bdf0ee857a65312859).

SparkQA commented:
[Test build #42004 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42004/console) for   PR 8180 at commit [`bb5190f`](https://github.com/apache/spark/commit/bb5190f0804f83fd178960bdf0ee857a65312859).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42004/
Test FAILed.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42002/
Test FAILed.

mateiz committed:
Further improvements to tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42015 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42015/consoleFull) for   PR 8180 at commit [`9d23757`](https://github.com/apache/spark/commit/9d23757d613234f700c3f6a0dfff1bf8d468021d).

mateiz committed:
Improve comments

mateiz commented:
Alright, I think this is ready to review now. Changes made:
- Added more docs to DAGScheduler about how stages may be re-attempted
- Added tests on:
  - More complex lineage graphs with multiple map jobs and result jobs pending
  - Fetch failures in the above, which also lead to resubmitted stages
  - Executor failure during a stage, and late task completion messages
- Better handling of getting the MapOutputStatistics object -- now the DAGScheduler gets it at a point when it knows the map output tracker has outputs for all tasks, whereas before the tracker might lose some outputs by the time SparkContext.submitMapStage queries it.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42016 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42016/consoleFull) for   PR 8180 at commit [`082b134`](https://github.com/apache/spark/commit/082b13481bcbaf537e29538a3f1335f4eba7005c).

SparkQA commented:
**[Test build #42015 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42015/console)**     for PR 8180 at commit [`9d23757`](https://github.com/apache/spark/commit/9d23757d613234f700c3f6a0dfff1bf8d468021d)     after a configured wait of `250m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42015/
Test FAILed.

SparkQA commented:
**[Test build #42016 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42016/console)**     for PR 8180 at commit [`082b134`](https://github.com/apache/spark/commit/082b13481bcbaf537e29538a3f1335f4eba7005c)     after a configured wait of `250m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42016/
Test FAILed.

mateiz commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42032 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42032/consoleFull) for   PR 8180 at commit [`082b134`](https://github.com/apache/spark/commit/082b13481bcbaf537e29538a3f1335f4eba7005c).

SparkQA commented:
[Test build #42032 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42032/console) for   PR 8180 at commit [`082b134`](https://github.com/apache/spark/commit/082b13481bcbaf537e29538a3f1335f4eba7005c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42032/
Test PASSed.

mateiz commented:
@zsxwing / @squito can you take a second look at this?

ryan-williams commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -45,17 +45,65 @@ import org.apache.spark.storage.BlockManagerMessages.BlockManagerHeartbeat
  * Here's a checklist to use when making or reviewing changes to this class:
  *
+ *  - All data structures should be cleared when the jobs involving them end to avoid indefinite
+ *    accumulation of state in long-runnin programs.
```
nit: long-running

zsxwing commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -720,31 +843,82 @@ class DAGScheduler(
+    // If the whole stage has already finished, tell the listener and remove it
+    if (!finalStage.outputLocs.contains(Nil)) {
+      job.finished(0) = true
+      job.numFinished += 1
```
Just feel a bit weird here, although `job.finished` and `job.numFinished` is meaningless for a `submitMapStage`. Maybe it's better to add some comments here.

mateiz replied:
Alright, I'll refactor these into a markMapStageJobAsFinished method that explains it.

zsxwing commented on core/src/main/scala/org/apache/spark/scheduler/ActiveJob.scala:
```diff
@@ -23,18 +23,42 @@ import org.apache.spark.TaskContext
+ *
+ * Jobs are only tracked for "leaf" stages that clients directly submitted, through DAGScheduler's
+ * submitJob or submitMapStage methods. However, either type of job may cause the execution of
+ * may other earlier stages (for RDDs in the DAG it depends on), and multiple jobs may share some
```
nit: `may` is redundant

zsxwing commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -45,17 +45,65 @@ import org.apache.spark.storage.BlockManagerMessages.BlockManagerHeartbeat
+ * To recover from failures, the same stage might need to run multiple times, which are called
+ * "attempts". If the TaskScheduler reports that a task failed because a map output file from a
+ * previous stage was lost, the DAGScheduler resubmits that lost stage. This is detected through a
+ * through a CompletionEvent with FetchFailed, or an ExecutorLost event. The DAGScheduler will wait
```
nit: `through a` is duplicate

zsxwing commented:
LGTM except some nits.

mateiz committed:
Review comments

mateiz commented:
Thanks for the comments; I've made the fixes. Let me know if anyone else has other comments.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42385 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42385/consoleFull) for   PR 8180 at commit [`8bfce4d`](https://github.com/apache/spark/commit/8bfce4d674cc0aa88c67d4defff054fc41216681).

SparkQA commented:
[Test build #42385 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42385/console) for   PR 8180 at commit [`8bfce4d`](https://github.com/apache/spark/commit/8bfce4d674cc0aa88c67d4defff054fc41216681).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](`
  * `case class ExecutorLostFailure(execId: String, isNormalExit: Boolean = false)`
  * `class CoGroupedRDD[K: ClassTag](`
  * `class ShuffledRDD[K: ClassTag, V: ClassTag, C: ClassTag](`
  * `class ExecutorLossReason(val message: String) extends Serializable `
  * `case class ExecutorExited(exitCode: Int, isNormalExit: Boolean, reason: String)`
  * `  case class RemoveExecutor(executorId: String, reason: ExecutorLossReason)`
  * `  case class GetExecutorLossReason(executorId: String) extends CoarseGrainedClusterMessage`
  * `class MultilayerPerceptronClassifier(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class MultilayerPerceptronClassificationModel(JavaModel):`
  * `class MinMaxScaler(JavaEstimator, HasInputCol, HasOutputCol):`
  * `class MinMaxScalerModel(JavaModel):`
  * `class StringIndexer(JavaEstimator, HasInputCol, HasOutputCol, HasHandleInvalid):`
  * `        ("thresholds", "Thresholds in multi-class classification to adjust the probability of " +`
  * `class HasHandleInvalid(Params):`
  * `class HasElasticNetParam(Params):`
  * `class HasFitIntercept(Params):`
  * `class HasStandardization(Params):`
  * `class HasThresholds(Params):`
  * `    thresholds = Param(Params._dummy(), "thresholds", "Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values >= 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class' threshold.")`
  * `        self.thresholds = Param(self, "thresholds", "Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values >= 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class' threshold.")`
  * `case class Stddev(child: Expression) extends StddevAgg(child) `
  * `case class StddevPop(child: Expression) extends StddevAgg(child) `
  * `case class StddevSamp(child: Expression) extends StddevAgg(child) `
  * `abstract class StddevAgg(child: Expression) extends AlgebraicAggregate `
  * `abstract class StddevAgg1(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class Stddev(child: Expression) extends StddevAgg1(child) `
  * `case class StddevPop(child: Expression) extends StddevAgg1(child) `
  * `case class StddevSamp(child: Expression) extends StddevAgg1(child) `
  * `case class ComputePartialStd(child: Expression) extends UnaryExpression with AggregateExpression1 `
  * `case class ComputePartialStdFunction (`
  * `case class MergePartialStd(`
  * `case class MergePartialStdFunction(`
  * `case class StddevFunction(`
  * `case class ConvertToSafeNode(conf: SQLConf, child: LocalNode) extends UnaryLocalNode(conf) `
  * `case class ConvertToUnsafeNode(conf: SQLConf, child: LocalNode) extends UnaryLocalNode(conf) `
  * `case class FilterNode(conf: SQLConf, condition: Expression, child: LocalNode)`
  * `case class HashJoinNode(`
  * `case class IntersectNode(conf: SQLConf, left: LocalNode, right: LocalNode)`
  * `case class LimitNode(conf: SQLConf, limit: Int, child: LocalNode) extends UnaryLocalNode(conf) `
  * `abstract class LocalNode(conf: SQLConf) extends TreeNode[LocalNode] with Logging `
  * `abstract class LeafLocalNode(conf: SQLConf) extends LocalNode(conf) `
  * `abstract class UnaryLocalNode(conf: SQLConf) extends LocalNode(conf) `
  * `abstract class BinaryLocalNode(conf: SQLConf) extends LocalNode(conf) `
  * `case class ProjectNode(conf: SQLConf, projectList: Seq[NamedExpression], child: LocalNode)`
  * `case class SampleNode(`
  * `case class SeqScanNode(conf: SQLConf, output: Seq[Attribute], data: Seq[InternalRow])`
  * `case class TakeOrderedAndProjectNode(`
  * `case class UnionNode(conf: SQLConf, children: Seq[LocalNode]) extends LocalNode(conf) `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42385/
Test PASSed.
